src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,484 @@
+      return node->fields[index];
+    }
+    const auto* sinfo = GetStructInfoAs<TupleStructInfoNode>(expr);
+    ICHECK(sinfo != nullptr) << "GetField: expr must has TupleStructInfo.";
+    Expr ret = TupleGetItem(expr, index);
+    UpdateStructInfo(ret, sinfo->fields[index]);
+    return ret;
+  }
+
+  AdjointMsg ExprToAdjointMsg(Expr expr) {
+    return MapToNestedMsg<Expr>(expr, GetField, [](Expr leaf) {
+      ICHECK(GetStructInfoAs<TensorStructInfoNode>(leaf))
+          << "The leaf of adjoint: " << leaf << " should have StructInfo and be a Tensor.";
+      return AdjointMsg(leaf);
+    });
+  }
+
+  // Transform the adjoint expressed as NestedMsg<Expr> into adjoint Expr, and then emit it
+  // If the adjoint is assigned to a DataflowVar (the adjoint corresponds to a non-output binding),
+  // it would be stored in adjoint_var_map_ for future lookup
+  Var EmitAdjoint(const Var& source_var, const AdjointMsg& adjoint, bool is_dataflow_var) {
+    Var adjoint_var;
+    if (is_dataflow_var) {
+      adjoint_var = builder_->Emit(AdjointMsgToExpr(adjoint), source_var->name_hint() + "_adjoint");
+      adjoint_var_map_.Set(source_var, adjoint_var);
+    } else {
+      adjoint_var =
+          builder_->EmitOutput(AdjointMsgToExpr(adjoint), source_var->name_hint() + "_adjoint");
+    }
+    return adjoint_var;
+  }
+
+  // Handle the return value of the AD function.
+  // Returns the new return value, which would be like:
+  // Tuple(original_return_value,
+  //       Tuple(adjoint_of_require_grads_1, adjoint_of_require_grads_2, ...))
+  Expr Epilogue(const Array<Var>& require_grads, const Var& target_var) {
+    // create adjoint variables for inputs, and then bind adjoints
+    Array<Expr> out_adjoints;
+
+    for (Var var : require_grads) {
+      // If the var don't have adjoint msg, it do not contribute to the target
+      // so its adjoint is zeros
+      AdjointMsg adjoint =
+          adjoint_msg_map_.Get(var).value_or(BuildZerosAdjointNested(GetStructInfo(var)));
+      Var adjoint_var = EmitAdjoint(var, adjoint, false);
+      out_adjoints.push_back(adjoint_var);
+    }
+
+    return Tuple(Array<Expr>{target_var, Tuple(out_adjoints)});
+  }
+
+  // The block builder of the corresponding GradientMutator, to emit bindings
+  BlockBuilder builder_;
+
+  // Forward Var to its adjoint Var
+  Map<Var, Var> adjoint_var_map_;
+  // Forward Var to its adjoint NestedMsg<Expr>
+  // Here we use NestedMsg<Expr> to save the adjoint information (equaivalent to adjoint Expr)
+  // When emiting, adjoint information will be transformed into adjoint Expr
+  Map<Var, AdjointMsg> adjoint_msg_map_;
+};
+
+class GradientMutator : public ExprMutator {
+ public:
+  explicit GradientMutator(const IRModule& mod, const GlobalVar& gvar,
+                           const Array<Var>& require_grads)
+      : ExprMutator(mod),
+        mod_(mod),
+        gvar_(std::move(gvar)),
+        require_grads_(std::move(require_grads)) {}
+
+  IRModule Transform() {
```

Comment:
Consider making the exposed interface a static function

Reply:
Should `BackwardBindingGenerator().Generate()` also be a static function?